Feature: Occurrences Widget Implementation

TomerPacific opened the pull request:
Adding the implementation for occurrences widget. This is a stateless widget which accepts three DateTime objects (which may be null). We display the content using a container with a column containing two rows

If any of the DateTime objects is null, a "--" will be displayed (see check in line 42 as an example).
There could be a better way to hide this if logic, but it becomes a bit more complicated as rendering widgets does not allow if/ternary logic. 

Resolves #22 .

![app](https://user-images.githubusercontent.com/23402988/66707414-b153d400-ed48-11e9-8857-f7c939c87ac0.jpg)

If there is a requirement for comments, please let me know.

TomerPacific committed:
Adding the impementation for occurances widget. This is a stateless widget which accepts three DateTime objects (which may be null). We display the content using a container with a column containing two rows

TomerPacific committed:
Changing colors for last updated text

Vanethos commented on lib/presentation/ui/occurrences_time_widget.dart:
```diff
@@ -0,0 +1,122 @@
+                                  ),
+                                ),
+                                new Text(
+                                  this._startTime != null ? this._startTime.hour.toString() + ":" + this._startTime.minute.toString() : "--",
```
You are using this same line of code twice.

Can you extract a method for it?

Also, consider using Date Formatting as seen here:
https://stackoverflow.com/a/16126580/4499889

Vanethos replied:
Also, to make it more readable, consider using `( ` and `)` on your logic

So:

`this._startTime != null ? (this._startTime.hour.toString() + ":" + this._startTime.minute.toString()) : "--",`

TomerPacific replied:
@Vanethos , please look at my updated PR

Vanethos commented on lib/presentation/ui/occurrences_time_widget.dart:
```diff
@@ -0,0 +1,122 @@
+                                ),
+                                new Text(
+                                  this._startTime != null ? this._startTime.day.toString() + "-" + this._startTime.month.toString() + "-" + this._startTime.year.toString() : "--",
+                                  overflow: TextOverflow.clip,
```
Why are you clipping the datetime here?

Wouldn't it be best for the text to be flexible and, if needed, go to the the next line so that the user can see the correct time?

You might consider wrapping the Text widget inside another widget such as a Container to wrap the text

TomerPacific replied:
@Vanethos , please look at my updated PR

Vanethos commented on lib/presentation/ui/occurrences_time_widget.dart:
```diff
@@ -0,0 +1,122 @@
+                          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+                          children: <Widget>[
+                            new Text(
+                        this._lastUpdated != null ?
```
As with the previous statement, if you could extract a method and use date formatting it would be awesome.

This way the widget tree would not have this kind of logic

TomerPacific replied:
@Vanethos , please look at my updated PR

Vanethos reviewed (changes requested):
Please review these comments.

Overall it was great! So congrats on the PR!

TomerPacific committed:
Refactoring date formatting logic to method getFormattedDate

TomerPacific committed:
Refactored code to use DateFormatter

TomerPacific committed:
Refactored getFormattedDate to allow to be used by all DateTime objects (start, end and lastUpdated)

TomerPacific committed:
Correcting hour format

TomerPacific committed:
Removing overflow attribute and wrapping widgets in a container widget

Vanethos commented:
I've tried the following code:

```dart
ListView(
        children: [
          OccurrencesTimeWidget(
              DateTime.now(), null, DateTime.now()
          ),
          OccurrencesTimeWidget(
              DateTime.now(), null, DateTime.now()
          ),
          OccurrencesTimeWidget(
              DateTime.now(), null, DateTime.now()
          ),
          OccurrencesTimeWidget(
              DateTime.now(), null, DateTime.now()
          ),
        ]
      )
```

And I had the following result:

<img width="453" alt="imagem" src="https://user-images.githubusercontent.com/10728633/66776481-7af88f00-eed7-11e9-8c05-df3a4e591210.png">


Also, when using just as a standalone widget in the page, I got this:

```dart
Container(
        child: Center(
          child: OccurrencesTimeWidget(
              DateTime.now(), null, DateTime.now()
          ),
        ),
      )
```

<img width="458" alt="imagem" src="https://user-images.githubusercontent.com/10728633/66776515-9d8aa800-eed7-11e9-95ae-0d5f963c4abd.png">


While the expected result is:

![imagem](https://user-images.githubusercontent.com/10728633/66776564-b5622c00-eed7-11e9-80da-bd367eb5c74d.png)

The size should be constrained so that we can use it isolated or in a list

TomerPacific commented:
@Vanethos ,
it is weird that you don't see anything in the second photograph and I don't understand why. Regarding the listview, you say that the size should be constrained. Would you rather the widget be wrapped inside of a [SizedBox ](https://api.flutter.dev/flutter/widgets/SizedBox-class.html)instead of a Container?

Vanethos commented:
@TomerPacific 

If you can, try to create the layout in a way that you don't need a sized box. The size of the widget should be equal to the text size plus the margins, so wrapping a container and setting the correct text size and the correct margins should just do it.

I suspect that this is because of the `fontSizeFactor` attribute. In higher DPI phones like the iPhone XR you may have strange behaviour.

If I remove that attribute, I can print the following:

<img width="447" alt="imagem" src="https://user-images.githubusercontent.com/10728633/66778733-a41c1e00-eedd-11e9-883f-6a3b9e3125de.png">

Though, as you can see, the text styles are different from the image.

Also the background color is different, you can try a lighter color.

As for the text not appearing, this is because the widget was not wrapped in a `SafeArea`, and since the text size was small, all the content was behind the notch :P 

I suggest that for now you use a `TextStyle` for that with a specific dimension for text size, for example

```
TextStyle styleHyperLinkedText() => TextStyle(
    fontSize: 14,
    color: Colors.grey,
    decoration: TextDecoration.underline);
```

TomerPacific committed:
Replacing style of Text widgets to TextStyle and correcting size and color

TomerPacific commented:
@Vanethos , let me know how it looks like now. I have updated the styles of the Text widgets and corrected the size and color.  All the text widgets are wrapped inside of a container widget.

![app](https://user-images.githubusercontent.com/23402988/66946269-3e54a100-f059-11e9-9487-e351588d9a3d.jpg)

Vanethos commented on lib/presentation/ui/occurrences_time_widget.dart:
```diff
@@ -0,0 +1,158 @@
+                ),
+                child:
+                    new Column(
+                      children: <Widget>[
```
If you put here: `mainAxisSize: MainAxisSize.min,` then the column will wrap the widget nicely

TomerPacific replied:
Added the attribute

Vanethos commented on lib/presentation/ui/occurrences_time_widget.dart:
```diff
@@ -0,0 +1,158 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return Material(
```
Why do you need the material ancestor?

TomerPacific replied:
I read and observed that wrapping the widgets with the Material widgets causes the yellow lines that appear on the Text widgets to disappear.

Without the Material ancestor:
![app](https://user-images.githubusercontent.com/23402988/67149809-6a219200-f2b8-11e9-8efe-0d17a8138a46.jpg)

If you want, I can commit the file with it removed.

Vanethos reviewed (commented):
Last comments

Vanethos commented:
@TomerPacific Last comments!

So, if you add `mainAxisSize: MainAxisSize.min,` to that widget, it will make it wrap it nicely when put on a list:

<img width="369" alt="imagem" src="https://user-images.githubusercontent.com/10728633/67108255-2a3aac00-f1df-11e9-98ea-819dad578767.png">


We can do it both ways:

If you don't have the time, I can approve this and add it on my own, no problem!

Or, if you want, you can add it to test it yourself

Your call :D

TomerPacific commented:
@Vanethos , please see my replies to your comments.

Vanethos commented:
@TomerPacific awesome 👍 

Going to approve your PR!

TomerPacific commented:
@Vanethos , pay attention that I did not make a commit to add the MainAxisSize.min as I was awaiting your input.

Thanks for your patience and feedback with this feature. 🙏

Vanethos commented:
Hi Tomer!

Yes, I saw that, but since it was such a small commit, I decided to approve your PR!

Thank YOU for your time and for helping out in our project! Your PR was greatly appreciated 👍
